keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,172 @@
+* Streamline the tedious processes involved in debugging latency issues in Kubernetes
+* Make it possible to identify high-level latency regressions, and attribute them back to offending processes
+
+
+### Non-Goals
+
+* To replace existing logging, metrics, or the events API
+* To trace operations from all Kubernetes resource types in a generic manner (i.e. without manual instrumentation)
+
+
+## Proposal
+
+### Architecture
+
+#### Trace lifecycle
+
+Kubernetes is unique in that it is constantly reconciling its actual state towards some desired state. As a result, it has no definitive concept of an "operation", which breaks the traditional model for distributed tracing. This raises the question of when to begin traces, and when to end them.
+
+In this proposal, we choose to _only_ trace phases of an object's lifecycle wherein it's correcting from an undesired state to its desired state, and to end the trace when it enters this desired state. This means that the same object will export traces for each reconciliation it undergoes. This decision was made because:
+
+1) These reconciliation phases are where Kubernetes performs actual work that can be traced
+2) Prevents traces from spanning the entire life of an object, which can be indefninite
+3) Provides a consistent, concrete duration for the root span which remains applicable across object types
+4) The time required for Kubernetes controllers to reconcile desired and actual state is, by definition, a measurement of either latency experienced by end-users in the fulfillment of their requests, or down-time due to a disruption
+
+#### Root spans
+
+In the standard model for distributed tracing, there exists a span in each trace that all other spans are descendents of and which extends the length of the entire trace, called the `root span`. 
+
+The Kubernetes component that kicks off an operation might not be the same component that ends it. In this proposal, when we are at the point where we want to end a root span, we craft a span to export which acts as the root span for the trace. For example, when the kubelet updates a pod from `Pending` to `Running`, it creates a root span using the start time of the pod as the start, and the current time as the end.
+
+
+#### Context propagation
+
+To correlate work done between components as belonging to the same trace, we must pass span context across process boundaries. In traditional distributed systems, this context can be passed down through RPC metadata or HTTP headers. Kubernetes, however, due to its watch-based nature, requires us to attach trace context directly to the target object. 
+
+In this proposal, we choose to propagate this span context as an encoded string an object annotation called `trace.kubernetes.io/context`. This annotation value is regenerated and replaced when an object's trace ends, to achieve the desired behavior from [section one](#trace-lifecycle). 
```

Comment:
I might be missing something, but this seems like it's prone to multiple "traces being started" concurrently causing race conditions where trace contexts are concurrently overwritten.

Reply:
As long as we ensure that there's a *single* state transition that we consider the beginning of a trace, and a *single* state transition that marks its end, I believe we should be able to avoid any race conditions here. 

@dashpole on this as well.

Reply:
It is worth noting that updates to an object's trace annotation should only be done by a single component, usually the controller responsible for updating the status of the object.  For example, the kubelet updates the annotation after updating the pod from pending -> running.

Reply:
On more thought, I think I understand where this is coming from.  Concurrent updates shouldn't be an issue, as the last update should be the trace context used, but there could be a race between "ending a trace" by replacing the trace context, and "starting a trace" from an update, for example.